sdk/attestation/Azure.Security.Attestation/src/Models/TokenSigningKey.cs

```diff
@@ -0,0 +1,109 @@
+        /// Creates a new instance of a <see cref="TokenSigningKey"/>
+        /// </summary>
+        /// <param name="signer">The Signing Key which will be used to sign the generated token.</param>
+        /// <param name="certificate">An X.509 Certificate which wraps the public key part of the <paramref name="signer"/>.</param>
+        public TokenSigningKey(AsymmetricAlgorithm signer, X509Certificate2 certificate)
+        {
+            Argument.AssertNotNull(signer, nameof(signer));
+            Argument.AssertNotNull(certificate, nameof(certificate));
+            Signer = signer;
+            Certificate = certificate;
+            VerifySignerMatchesCertificate(signer, certificate);
+        }
+
+        /// <summary>
+        /// Creates a new instance of a <see cref="TokenSigningKey"/>
+        /// </summary>
+        /// <param name="certificate">An X.509 Certificate with a private key.</param>
+        public TokenSigningKey(X509Certificate2 certificate)
+        {
+            Argument.AssertNotNull(certificate, nameof(certificate));
+            if (!certificate.HasPrivateKey)
+            {
+                throw new ArgumentException($"Certificate provided {certificate.ToString()} does not have a private key.");
+            }
+            Signer = certificate.PrivateKey;
+            Certificate = certificate;
+            VerifySignerMatchesCertificate(Signer, Certificate);
+        }
+
+        /// <summary>
+        /// Gets the signer used to sign the attestation token
+        /// </summary>
+        public AsymmetricAlgorithm Signer { get; }
+
+        /// <summary>
+        /// Gets the X.509 Certificate which will be transmitted to the attestation service to validate the signed token.
+        /// </summary>
+        public X509Certificate2 Certificate { get; }
+
+        private static void VerifySignerMatchesCertificate(AsymmetricAlgorithm signer, X509Certificate2 certificate)
+        {
+            if (!signer.KeyExchangeAlgorithm.StartsWith(certificate.PublicKey.Key.KeyExchangeAlgorithm, System.StringComparison.Ordinal))
+            {
+                throw new ArgumentException($"Signer key algorithm {signer.SignatureAlgorithm} does not match certificate key algorithm {certificate.PublicKey.Key.SignatureAlgorithm}");
+            }
+
+            if (signer is RSA rsaKey)
+            {
+                var certificateKey = certificate.PublicKey.Key as RSA;
+                RSAParameters signerParameters = rsaKey.ExportParameters(false);
+                RSAParameters certificateParameters = certificateKey.ExportParameters(false);
+                if (!signerParameters.Modulus.SequenceEqual(certificateParameters.Modulus) ||
+                    !signerParameters.Exponent.SequenceEqual(certificateParameters.Exponent))
+                {
+                    throw new ArgumentException($"Signer key {signer} does not match certificate key {certificateKey}");
+                }
+            }
+#if false
```

Comment:
Are we intentionally ignoring some branches, or was this potentially accidental from local testing?

Reply:
Accidental - ECDH fails badly on .net core. Let me see if I can come up with a better solution.

Reply:
Ok, this is a mess. My preferred solution is to use ToXmlString to compare two keys. However that doesn't work on NetCoreApp2.1 builds (it does everywhere else).

Any suggestions about how to conditionally compile the code to make this work?

Reply:
Would it be possible to somehow perform a trial encryption/decryption?
* First encrypt with signer
* Second decrypt with certificate.publickey

Getting the public key as a cryptographic entity seems like it might be painful, but the [docs](https://docs.microsoft.com/en-us/dotnet/api/system.security.cryptography.x509certificates.x509certificate2.publickey?view=net-5.0) say this:

You can also obtain the key as an AsymmetricAlgorithm object by referencing the PublicKey property. This property supports only RSA or DSA keys, so it returns either an RSACryptoServiceProvider or a DSACryptoServiceProvider object that represents the public key.

Reply:
With respect to conditional compilation, there's no definitive answer.  It's pretty much what you'd expect.   Most often, we'll use the same `#if` approach as we are here with the applicable [framework preprocessor symbols](https://docs.microsoft.com/dotnet/standard/frameworks).

For strange cases, we may do conditional `<Compile>` inclusions in the project file using some of the build property definitions as the condition - but this tends to be somewhat difficult to discover and understand, quite a pain to manage, and, in my experience, rarely used.

Reply:
@jsquire: I chased down the examples in the tree, most of the conditionals are there only for test collateral, not for the actual code. I kinda like Greg's sign/verify suggestion, but it may have a significant negative performance impact. On the other hand, the expectation is that you're either about to hit the wire for the operation, or you have just hit the wire for the operation, so that isn't likely to be a huge deal.

Reply:
@gkostal , I tried your suggestion and it works like a champ. Thanks.